parlai/agents/vsepp/vsepp.py

```diff
@@ -0,0 +1,281 @@
+            loss, top1, ranks = self.predict(images, xs, x_lens, cands=None,
+                                             is_training=is_training)
+            if loss is not None:
+                batch_reply[0]['metrics'] = {'loss': loss.item()}
+            predictions = []
+            for score_idx in top1:
+                predictions.append(labels[score_idx])
+        else:
+            # Need to collate then sort the captions by length
+            cands = [self.candidate_helper(vec_obs[idx]['candidate_labels_vec'],
+                                           vec_obs[idx]['candidate_labels'],
+                                           is_testing)
+                     for idx in valid_inds]
+            _, top1, ranks = self.predict(images, None, None, cands,
+                                          is_training)
+            predictions = []
+            for i, score_idx in enumerate(top1):
+                predictions.append(cands[i][1][score_idx])
+
+        unmap_ranks = self.unmap_valid(ranks, valid_inds, batch_size)
+        unmap_pred = self.unmap_valid(predictions, valid_inds, batch_size)
+
+        for i, (rep, pred) in enumerate(zip(batch_reply, unmap_pred)):
+            if pred is not None:
+                rep['text'] = pred
+                if not is_testing:
+                    rep['truth_rank'] = unmap_ranks[i]
+
+        return batch_reply
+
+    def predict(self, xs, ys=None, y_lens=None, cands=None, is_training=False):
+        loss = None
+        if is_training:
+            self.model.train()
+            self.optimizer.zero_grad()
+            img_embs, cap_embs = self.model(xs, ys, y_lens)
+            loss, ranks, top1 = self.criterion(img_embs, cap_embs)
+            self.metrics['loss'] += loss.item()
+            self.metrics['r@'] += ranks
+            loss.backward()
+            self.optimizer.step()
+        else:
+            self.model.eval()
+            # Obtain the image embeddings
+            img_embs, _ = self.model(xs, None, None)
+            ranks = []
+            top1 = []
+            # Each image has their own caption candidates, so we need to
+            # iteratively create the embeddings and rank
+            for i, (cap, _, lens, truth_idx) in enumerate(cands):
+                _, embs = self.model(None, cap, lens)
+                # Hack to pass through the truth label's index to compute the
+                # rank and top metrics
+                offset = truth_idx if truth_idx is not None else 0
+                _, rank, top = self.criterion(img_embs[i, :].unsqueeze(0),
+                                              embs, offset)
+                ranks += rank
+                top1.append(top[0])
+            self.metrics['r@'] += ranks
+
+        return loss, top1, ranks
+
+    def report(self):
+        m = {}
+        m['loss'] = self.metrics['loss']
+        ranks = np.asarray(self.metrics['r@'])
+        m['r@1'] = len(np.where(ranks < 1)[0]) / len(ranks)*1.0
```

Comment:
i'm not sure but it looks like this `*1.0` is an artifact from python 2 - `/` by default returns floats now